javascript/ql/src/semmle/javascript/PrintAst.qll

```diff
@@ -0,0 +1,808 @@
+/**
+ * Provides queries to pretty-print a JavaScript AST as a graph.
+ *
+ * By default, this will print the AST for all elements in the database. To change this behavior,
+ * extend `PrintAstConfiguration` and override `shouldPrint` to hold for only the elements
+ * you wish to view the AST for.
+ */
+
+import javascript
+
+private newtype TPrintAstConfiguration = MkPrintAstConfiguration()
+
+/**
+ * The query can extend this class to control which elements are printed.
+ */
+class PrintAstConfiguration extends TPrintAstConfiguration {
+  /**
+   * Gets a textual representation of this `PrintAstConfiguration`.
+   */
+  string toString() { result = "PrintAstConfiguration" }
+
+  /**
+   * Controls whether the `Element` should be considered for AST printing.
+   * By default it checks whether the `Element` `e` belongs to `Location` `l`.
+   */
+  predicate shouldPrint(Locatable e, Location l) { l = e.getLocation() }
+}
+
+private predicate shouldPrint(Locatable e, Location l) {
+  exists(PrintAstConfiguration config | config.shouldPrint(e, l))
+}
+
+/** Holds if the given element does not need to be rendered in the AST, due to being compiler-generated or being a `TopLevel`. */
+private predicate isNotNeeded(Locatable el) {
+  exists(ClassDefinition c, ConstructorDeclaration constructor |
+    constructor = c.getConstructor() and
+    constructor.isSynthetic() and
+    el = constructor
```

Comment:
Wouldn't it be a good idea to render the implicit constructor anyway? It's part of the AST.